### PR TITLE
Use markdown-it and katex

### PR DESCRIPTION
**Pull request by billti**
This addresses #845 

Previously we used 'marked' and 'mathjax' in the playground, but 'markdown-it' and 'katex' in VS Code - the latter being what VS Code uses for its Markdown and LaTeX rendering in file previews, Jupyter notebooks, etc. This converts the playground, and the kata content generation, to use markdown-it and katex, and removes 'marked' and 'mathjax' from the repo.

Note that KaTeX is more strict that MathJax that the LaTeX is well formed. This is probably a _good thing_ while we're authoring content. I did need to clean up quite a bit of LaTeX in testing this, and there's probably quite a bit more elsewhere to go over. (In general our LaTeX is often in poor shape. We might want to look into any 'linting' options for it). Feel free to skip the .md files and focus on the code initially.

I also significantly revamped how we render markdown in a Preact components. This has led to much cleaner code, and deleting a lot of useRef and useEffect and dangerouslySetInnerHtml code, and the need to re-typeset content. (Esp. see kata.tsx and docs.tsx).

### CODE REVIEW

File: npm/qsharp/ux/renderers.tsx [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  const err = "ERROR: Rendered has not been set";
+  console.error(err);
+  return err + ". " + input;
+};
+
+export function setRenderer(renderer: (input: string) => string) {
+  the_renderer = renderer;
+}
+
+export function RenderDiv(props: { input: string; className?: string }) {
```
Comment (DmitryVasilevsky) [inline npm/qsharp/ux/renderers.tsx:16]: Consider RenderedDiv and RenderedLi names instead. They will be used like other elements such as table, span, etc.

**Review by DmitryVasilevsky — APPROVED** [review]

**Review by minestarks — APPROVED** [review]
:shipit:

File: katas/content/multi_qubit_gates/index.md [hunk 2]
```diff
@@ -304,7 +304,7 @@ $$=|00\rangle\langle00| + |01\rangle\langle01| + |10\rangle\langle11| + |11\rang
     "title": "SWAP Gate"
 })
 
-The $SWAP$ gate acts on two qubits, and, as the name implies, swaps their quantum states.
+The $\text{SWAP}$ gate acts on two qubits, and, as the name implies, swaps their quantum states.
```
Comment (tcNickolas) [inline katas/content/multi_qubit_gates/index.md:307]: Let's keep this as $SWAP$ for now for consistency, since it's by far not the only occurrence of a multi-letter gate in the katas. Once we agree on the style we should use for gates, we can do the update more broadly

File: katas/content/multi_qubit_gates/preparing_bell_state/index.md [hunk 3]
```diff
@@ -1,3 +1,3 @@
 **Input:** Two qubits in state $|00\rangle$, stored in an array of length 2.
 
-**Goal:** Transform the system into the Bell state $\ket{\Phi^+} = \frac{1}{\sqrt{2}}\big(|00\rangle + |11\rangle\big)$.
```
Comment (tcNickolas) [inline katas/content/multi_qubit_gates/preparing_bell_state/index.md:3]: What was the change supposed to be here? There was a conflict with my fix to add \ket to \Phi, and I couldn't figure out what was the proposed change

Comment (billti) [reply katas/content/multi_qubit_gates/preparing_bell_state/index.md:3]: The only change is a newline added to the end of the file.

File: katas/content/multi_qubit_gates/index.md (hunk 2 above)
Comment (billti) [reply katas/content/multi_qubit_gates/index.md:307]: Done in incoming changes.

File: katas/content/qubit/index.md [hunk 4]
```diff
@@ -44,50 +44,50 @@ $\alpha$ and $\beta$ are known as the probability amplitudes of states $0$ and $
...
+\langle \begin{bmatrix} 0 \\ 1 \end{bmatrix} , \begin{bmatrix} 0 \\ 1 \end{bmatrix} \rangle = 1
 $$
 
 2. They are orthogonal to each other.
 
 $$
-\langle \begin{bmatrix} 1 \\\ 0 \end{bmatrix} , \begin{bmatrix} 0 \\\ 1 \end{bmatrix} \rangle =
-\langle \begin{bmatrix} 0 \\\ 1 \end{bmatrix} , \begin{bmatrix} 1 \\\ 0 \end{bmatrix} \\rangle = 0
+\langle \begin{bmatrix} 1 \\ 0 \end{bmatrix} , \begin{bmatrix} 0 \\ 1 \end{bmatrix} \rangle =
+\langle \begin{bmatrix} 0 \\ 1 \end{bmatrix} , \begin{bmatrix} 1 \\ 0 \end{bmatrix} \\rangle = 0
```
Comment (tcNickolas) [inline katas/content/qubit/index.md:74]:
```suggestion
\langle \begin{bmatrix} 0 \\ 1 \end{bmatrix} , \begin{bmatrix} 1 \\ 0 \end{bmatrix} \rangle = 0
```

File: katas/content/qubit/index.md [hunk 5]
```diff
@@ -44,50 +44,50 @@ $\alpha$ and $\beta$ are known as the probability amplitudes of states $0$ and $
...
-Note that you can use scalar multiplication and vector addition to express any qubit state $\begin{bmatrix} \alpha \\\ \beta \end{bmatrix}$ as a sum of these two vectors with certain weights $\alpha$ and $\beta$, known as linear combination.
+Note that you can use scalar multiplication and vector addition to express any qubit state $\begin{bmatrix} \alpha \\ \beta \end{bmatrix}$ as a sum of these two vectors with certain weights $\alpha$ and $\beta$, known as linear combination.
 
 $$
-\begin{bmatrix} \alpha \\\ \beta \\end{bmatrix} =
-\begin{bmatrix} \alpha \\\ 0 \end{bmatrix} + \begin{bmatrix} 0 \\\ \beta \\end{bmatrix} =
-\alpha \cdot \begin{bmatrix} 1 \\\ 0 \end{bmatrix} + \beta \cdot \begin{bmatrix} 0 \\\ 1 \end{bmatrix}
+\begin{bmatrix} \alpha \\ \beta \\end{bmatrix} =
+\begin{bmatrix} \alpha \\ 0 \end{bmatrix} + \begin{bmatrix} 0 \\ \beta \\end{bmatrix} =
+\alpha \cdot \begin{bmatrix} 1 \\ 0 \end{bmatrix} + \beta \cdot \begin{bmatrix} 0 \\ 1 \end{bmatrix}
```
Comment (tcNickolas) [inline katas/content/qubit/index.md:58]: This formula errors out, I think it should be this fix, but haven't tried
```suggestion
\begin{bmatrix} \alpha \\ \beta \end{bmatrix} =
\begin{bmatrix} \alpha \\ 0 \end{bmatrix} + \begin{bmatrix} 0 \\ \beta \end{bmatrix} =
\alpha \cdot \begin{bmatrix} 1 \\ 0 \end{bmatrix} + \beta \cdot \begin{bmatrix} 0 \\ 1 \end{bmatrix}
```

File: katas/content/qubit/index.md [hunk 6]
```diff
@@ -134,13 +134,13 @@ We will learn more about Dirac notation in the next katas, as we introduce quant
...
 > between $-\pi$ and $\pi$ that has $\cos \theta = x$ and $\sin \theta = y$. Unlike using $\tan^{-1}(\frac{y}{x})$, `atan2` computes
 > the correct quadrant for the angle, since it preserves information about the signs of both sine and cosine of the angle.
 
-The probability amplitudes $\alpha$ and $\beta$ are complex numbers, therefore $\alpha$ and $\beta$ have a phase. For example, consider a qubit in state $\frac{1 + i}{2}|0\rangle + \frac{1 - i}{2}|1\rangle$. If you do the math, you see that the phase of $|0\rangle$ is $\atan2(\frac12, \frac12) = \frac{\pi}{4}$, and the phase of $|1\rangle$ is $\atan2(\frac12, -\frac12) = -\frac{\pi}{4}$. The difference between these two phases is known as **relative phase**.
+The probability amplitudes $\alpha$ and $\beta$ are complex numbers, therefore $\alpha$ and $\beta$ have a phase. For example, consider a qubit in state $\frac{1 + i}{2}|0\rangle + \frac{1 - i}{2}|1\rangle$. If you do the math, you see that the phase of $|0\rangle$ is $atan2(\frac12, \frac12) = \frac{\pi}{4}$, and the phase of $|1\rangle$ is $atan2(\frac12, -\frac12) = -\frac{\pi}{4}$. The difference between these two phases is known as **relative phase**.
```
Comment (tcNickolas) [inline katas/content/qubit/index.md:143]: \atan2 here is deliberate, to use the same spelling as trigonometric functions

**Review by tcNickolas — CHANGES_REQUESTED** [review]
I'll post my comments in batches, I'm going through all katas and not sure when I'll finish the last one. Here are comments for the first three, Complex Arithmetic, Linear Algebra, and Qubit.

(There are some things we need to fix independently, which I won't include to keep the scope somewhat reasonable :-) For example, ComplexArithmetic "Powers of $i$" doesn't render, but I'll be renaming exercises anyways and can change it to "Powers of imaginary unit")

**Review by tcNickolas — CHANGES_REQUESTED** [review]
Review for Single-Qubit Gates, Multi-Qubit Systems and Gates, and Superposition

Also, in the "Even and Odd Numbers" task, the Markdown tables don't render any longer, can you update them same as you did in Complex Arithmetic? It would be nice to be able to render Markdown tables too, since they are a lot easier to author than HTML ones

File: katas/content/qubit/index.md (hunk 6 above)
Comment (billti) [reply katas/content/qubit/index.md:143]: But \atan2 isn't a valid latex command (in MathJax or Katex) and gives errors when used. (\arctan is, but that's not what you want).

Comment (tcNickolas) [reply katas/content/qubit/index.md:143]: Right now on QCOM \atan2 renders successfully:
image

Another discrepancy that we need to track down and fix on QCOM?

Comment (billti) [reply katas/content/qubit/index.md:143]: I think QCOM just has MathJax set to 'ignore all errors' mode. If you go to the MathJax demo page at https://www.mathjax.org/#demo and put it in you'll see it's highlighted as an error.

File: katas/content/superposition/bell_state/solution.md [hunk 7]
```diff
@@ -33,18 +33,18 @@ Steps required to reach goal state:
...
-\underset{|0\rangle}{\underbrace{ \begin{bmatrix} 1 \\\ 0 \\\ 0 \\\ 0 \end{bmatrix}}}
-= \frac{1}{\sqrt2} \begin{bmatrix} 1 & 0 & 1 & 0 \\\ 0 & 1 & 0 & 1 \\\ 0 & 1 & 0 & -1 \\\ 1 & 0 & -1 & 0 \end{bmatrix}
+\underset{|0\rangle}{\underbrace{ \begin{bmatrix} 1 \\ 0 \\ 0 \\ 0 \end{bmatrix}}}
+= \frac{1}{\sqrt2} \begin{bmatrix} 1 & 0 & 1 & 0 \\ 0 & 1 & 0 & 1 \\ 0 & 1 & 0 & -1 \\ 1 & 0 & -1 & 0 \end{bmatrix}
 \cdot
-\begin{bmatrix} 1 \\\ 0 \\\ 0 \\\ 0 \end{bmatrix}
-= \underset{goal}{\underbrace{ \frac{1}{\sqrt2} \begin{bmatrix} 1 \\\ 0 \\\ 0 \\\ 1 \end{bmatrix}}}
+\begin{bmatrix} 1 \\ 0 \\ 0 \\ 0 \end{bmatrix}
+= \underset{goal}{\underbrace{ \frac{1}{\sqrt2} \begin{bmatrix} 1 \\ 0 \\ 0 \\ 1 \end{bmatrix}}}
 \label{6.1} \tag{6.1}
```
[the remark this answers is not in the document]
Comment (billti) [reply katas/content/superposition/bell_state/solution.md:48]: `\tag` works fine. `\label` doesn't. What are these needed for?

[the remark this answers is not in the document]
Comment (tcNickolas) [reply katas/content/superposition/bell_state/solution.md:48]: Number equations and then reference them from later in the text

[the remark this answers is not in the document]
Comment (tcNickolas) [reply katas/content/superposition/bell_state/solution.md:48]: We discussed offline and decided to remove \label and \tag and rephrase the solutions to refer to those formulas in a different way. I'll do it in my next round of edits later this week, so we can resolve this conversation.

File: katas/content/superposition/even_sup_two_qubits_complex_phases/solution.md [hunk 8]
```diff
@@ -36,7 +36,7 @@ $$
```
[the remark this answers is not in the document]
Comment (tcNickolas) [reply katas/content/superposition/even_sup_two_qubits_complex_phases/solution.md:36]: Same here, resolving and will remove \label later

**Review by tcNickolas — APPROVED** [review]
I checked the katas rendering (in both views) building locally, and we also verified it in QCOM environment - looks good to me! Let's ship it, and then double check everything once we have a deployment in the test environment.

File: npm/qsharp/ux/renderers.tsx (hunk 1 above)
Comment (billti) [reply npm/qsharp/ux/renderers.tsx]: Changed.
